[Editor] Add the possibility to create a new editor in using the keyboard (bug 1853424)

calixteman opened the pull request:
When an editing button is disabled, focused and the user press Enter (or space), an editor is automatically added at the center of the current page. Next creations can be done in using the same keys within the focused page.

Snuffleupagus commented on src/display/editor/annotation_editor_layer.js:
```diff
@@ -484,6 +484,11 @@ class AnnotationEditorLayer {
     return editor;
   }
 
+  addNewEditor() {
+    const { width, height } = this.div.getBoundingClientRect();
+    this.#createAndAddNewEditor({ offsetX: width / 2, offsetY: height / 2 });
+  }
+
```
This can be removed, since we've already added such a method in the meantime; see https://github.com/mozilla/pdf.js/blob/c377f2d48860ce7c9aba9dbf1f6f4835762c7d77/src/display/editor/annotation_editor_layer.js#L553-L560

Snuffleupagus commented on src/display/editor/tools.js:
```diff
@@ -990,8 +998,10 @@ class AnnotationEditorUIManager {
    * Change the editor mode (None, FreeText, Ink, ...)
    * @param {number} mode
    * @param {string|null} editId
+   * @param {boolean} isFromKeyboard - true if the mode change is due to a
+   * keyboard action.
```
```suggestion
   * @param {boolean} [isFromKeyboard] - true if the mode change is due to a
   *   keyboard action.
```

Snuffleupagus commented on src/display/editor/tools.js:
```diff
@@ -990,8 +998,10 @@ class AnnotationEditorUIManager {
+   * keyboard action.
    */
-  updateMode(mode, editId = null) {
+  updateMode(mode, editId = null, isFromKeyboard) {
```
```suggestion
  updateMode(mode, editId = null, isFromKeyboard = false) {
```

Snuffleupagus commented on test/integration/freetext_editor_spec.js:
Please update the new test-cases to use the helpers that you introduced in PR #17036, such that we don't have to add all these arbitrary timeouts.

Snuffleupagus reviewed (changes requested):
Sorry about the delay, this PR slipped my mind.

calixteman committed:
[Editor] Add the possibility to create a new editor in using the keyboard (bug 1853424)

When an editing button is disabled, focused and the user press Enter (or space), an
editor is automatically added at the center of the current page.
Next creations can be done in using the same keys within the focused page.

Snuffleupagus commented:
It seems that this will make it *more difficult* for the user to *disable* editing with the keyboard:
 - Tab to the FreeText-button.
 - Press `Enter` to enable FreeText-editing.
 - Decide that you don't want to add any text and press `Escape`.
 - Try to *disable* FreeText-editing by pressing `Enter` again.
 - You're again prompted to add a new FreeText-annotation, with no good way to actually disable the feature.

Snuffleupagus commented on src/display/editor/tools.js:
```diff
@@ -1164,6 +1186,11 @@ class AnnotationEditorUIManager {
       layer.updateMode(mode);
     }
+    if (!editId && isFromKeyboard) {
+      this.addNewEditorFromKeyboard();
```
Given that the method consists of a single line, can't we just inline it here?

Snuffleupagus commented on src/display/editor/tools.js:
```diff
@@ -1432,6 +1463,13 @@ class AnnotationEditorUIManager {
+  get isEnterHandled() {
+    return (
+      this.#selectedEditors.size === 1 &&
+      [...this.#selectedEditors][0].isEnterHandled
```
We've now using two different ways of accessing the first element of this `Set`, note [this existing code](https://github.com/mozilla/pdf.js/blob/552cc059cc40fb3e5e7d6185c85a567821b6416c/src/display/editor/tools.js#L1777). How about introducing a common helper for that, e.g. something like the following?
```javascript
get firstSelectedEditor() {
  return this.#selectedEditors.keys().next().value;
}

calixteman commented:
> It seems that this will make it _more difficult_ for the user to _disable_ editing with the keyboard:
> 
>     * Tab to the FreeText-button.
> 
>     * Press `Enter` to enable FreeText-editing.
> 
>     * Decide that you don't want to add any text and press `Escape`.
> 
>     * Try to _disable_ FreeText-editing by pressing `Enter` again.
> 
>     * You're again prompted to add a new FreeText-annotation, with no good way to actually disable the feature.

In this case you need to move the focus back to the freetext button and hit enter.
What do you mean by _more difficult_ ? right now, the only way to disable the feature with the keyboard is move the focus on the button and then press enter. Am I missing something ?

Snuffleupagus commented:
> In this case you need to move the focus back to the freetext button and hit enter.

Sorry, then I apparently misunderstood how all this was supposed to work :-)

Snuffleupagus commented:
/botio-linux preview

Snuffleupagus reviewed (approved):
r=me, thanks!

calixteman commented:
/botio integrationtest
